File: Tree/c++/112_path_sum.cpp
#include<iostream>
#include<TreeNode.h>

using namespace std;

class Solution{
public:
    bool hasPathSum(TreeNode* root, int targetSum){
        if(!root){
            return false;
        }
        if(!root->left && !root->right && targetSum == root->val){
            return true;
        }
        return hasPathSum(root->left, targetSum - root->val) || hasPathSum(root->right, targetSum - root->val);
        //dps 每遍历一个节点减去该节点的val,当到叶子节点时如果sum为0则该路径存在
    }
}

File: Tree/python/112_path_sum.py
from TreeNode import TreeNode

class Solution:
    def hasPathSum(self, root : TreeNode, targerSum : int) -> bool:
        if not root:
            return False
        
        if not root.left and (not root.right) and targerSum == root.val:
            return True

        if root.left:
            targerSum = targerSum - root.val
            if self(root.left, targerSum):
                return True
            targerSum += root.val
        
        